packages/core/src/contrib/Constraints.ts

```diff
@@ -147,6 +149,26 @@ const constrDictSimple = {
 // -------- General constraints
 // Defined for all shapes, generally require shape queries or call multiple specific constraints.
 const constrDictGeneral = {
+  /** Require that `shape` is on the canvas */
+  onCanvas: (
+    [shapeType, props]: any,
+    canvasWidth: VarAD,
+    canvasHeight: VarAD
+  ) => {
+    const box = bboxFromShape([shapeType, props]);
+    const canvasXRange: [VarAD, VarAD] = [
+      neg(div(canvasWidth, constOf(2))),
+      div(canvasWidth, constOf(2)),
```

Comment:
Not sure whether this is true, but `mul(canvasWidth, constOf(0.5))` may be a bit more efficient?